### PR TITLE
[CUSPARSE] Add sv! and sm! routines

### PR DESCRIPTION
**Pull request by amontoison**
`sv2` and `sm2` will be removed in the next major version of CUDA toolkit.

### CODE REVIEW

**Comment by amontoison** [comment]
`ERROR: LoadError: CUDA error: initialization error (code 3, ERROR_NOT_INITIALIZED)`
It's hard to know the origin of the problem :thinking: 

Solved: The `if CUSPARSE.version() < v"11.5.1"` must be inside functions.

**Comment by amontoison** [comment]
@maleadt Do you idea what is the problem with these [lines](
https://github.com/JuliaGPU/CUDA.jl/pull/1593/files#diff-416606e9f5d8f57045482fc9fc855bccc8f31a27969e609160b17cc793501ae6R388-R389) ?

```julia
CUDA.CUSPARSE.cusparseSpMatGetAttribute(descA, CUSPARSE_SPMAT_FILL_MODE, 'L', Csize_t(4))

Internal error: encountered unexpected error during compilation of cusparseSpMatGetAttribute:
TypeError(func=:ccall, context="", expected=Symbol, got=CUDA.Deps.libcusparse)
...
ERROR: `ccall` requires the compiler
```

**Comment by maleadt** [comment]
There an issue with the generated headers, you need to use `libcusparse()` in the `ccall`.

**Comment by amontoison** [comment]
Thanks @maleadt! The tests passed with the last version of CUDA toolkit.
Should we just test with the last version in `test/cusparse/generic.jl`?

I have the same issue with #1592.

**Comment by maleadt** [comment]
What exactly is the issue? Can you find something relevant in the CUSPARSE changelogs? We can always make tests conditional, there's many like that already, but only if there's a good reason.

**Comment by amontoison** [comment]
We have a lot of issues related to `cusparseSpSV`, `cusparseSpSM`, `cusparseSpMV` and `cusparseSpMM` in the CUSPARSE changelogs:
```
cuSPARSE: Release 11.7 Update 1]
    New Features
- Improved COO cusparseSpMM Alg2 with support for batched computation, custom row/col-major layout for B/C, and mixed-precision computation.
- Better performance for cusparseSpMM COO Alg3 and cusparseSpSM.
    Resolved Issues
- Batched cusparseSpMM produced wrong results when the number of columns of B/C is one.
    Known Issues
- cusparseSpSV, cusparseSpSM could produce wrong results if the output vector/matrix is not zero-initialized.
```
```
[cuSPARSE: Release 11.7]
    Resolved Issues
- Fixed a rare correctness bug of cusparseSpMM with CUSPARSE_SPMM_CSR_ALG1 when the number of rows in the sparse matrix is 2.
- cusparseSpSV, cusparseSpSM could produce wrong results if the output vector/matrix is not zero-initialized.

```
```
cuSPARSE: Release 11.6 Update 1

    New Features
- Improved CSR cusparseSpMM Alg1:
- Improved COO cusparseSpMM Alg3 with support for batched computation, custom row/col-major layout for B/C, and mixed-precision computation.
- Improved mixed-precision computation of CSR/COO cusparseSpMV.
- Added CSC format support for cusparseSpMV and cusparseSpMM.
    Known Issues
- cusparseSpSV, cusparseSpSM could produce wrong results if the output vector/matrix is not zero-initialized.
```
```
cuSPARSE: Release 11.6
    Resolved Issues
- Fixed incorrect results in COO SpMM Alg1 which occurred in some rare cases.
```
```
cuSPARSE: Release 11.5 Update 1
    Resolved Issues
- cusparseSpSM produced wrong results in some cases when the matB operation is CUSPARSE_ OPERATION_NON_TRANSPOSE or CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE.
- cusparseSpSM produced wrong results in some cases when the matrix layout is row-major.
```
```
cuSPARSE: Release 11.4 Update 1
    Resolved Issues
- cusparseSpSV and cusparseSpSM could produce wrong results
- cusparseSpSV and cusparseSpSM did not work correctly when vecX == vecY or matB == matC.
```
```
cuSPARSE: Release 11.4
    Known Issues
- cusparseSpSV and cusparseSpSM could produce wrong results
- cusparseSpSV and cusparseSpSM do not work correctly when vecX == vecY or matB == matC.
```

I also submitted a bug to Nvidia yesterday:
- cusparseSpMV and cusparseSpMM: complex matrix-vector and matrix-matrix products with matrices stored CSC format and transa='C' return a wrong result (#1592) 
- cusparseSpSV and cusparseSpSM: forward and backward sweeps with complex triangular matrices, transa='C' and diag='N' return a wrong result (#1593)

**Comment by amontoison** [comment]
@maleadt 
I have feedbacks from the CUSPARSE engineers:
- cusparseMV isn't supporting conj transpose for matrix A in CSC format at the moment. We'll add this feature in the next release.
- For cusparseSpSV and SpSM kernels: cusparse supports CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE for CSR and COO. We found an issue that the diagonal of the sparse matrix A is not conjugated with CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE. This is why the problem only shows for non-unit diagonals. This is a known issue and we will fix it in the next release.

I propose to wait the next release for #1592.
For this PR, I modified the tests to take into account the issue with SpSV and SpSM but we can also wait the next release.

File: lib/cusparse/generic.jl [hunk 1]
```diff
@@ -57,11 +59,24 @@ Base.unsafe_convert(::Type{cusparseDnMatDescr_t}, desc::CuDenseMatrixDescriptor)
...
+        finalizer(cusparseDestroySpMat, obj)
+        return obj
+    end
+
     function CuSparseMatrixDescriptor(A::CuSparseMatrixCSR, IndexBase::Char)
         desc_ref = Ref{cusparseSpMatDescr_t}()
         cusparseCreateCsr(
             desc_ref,
-            size(A)..., length(nonzeros(A)),
+            size(A)..., A.nnz,
```
Comment (maleadt) [inline lib/cusparse/generic.jl]: calling `nnz` would be even better.

**Comment by maleadt** [comment]
Thanks for investigating!

>  For this PR, I modified the tests to take into account the issue with SpSV and SpSM but we can also wait the next release.

I'd be fine merging now, seeing how there's functionality/tests that are working already. Maybe add a note of open an issue to keep track of the functionality to enable for the next CUSPARSE release though?

**Comment by amontoison** [comment]
> Thanks for investigating!
> 
> > For this PR, I modified the tests to take into account the issue with SpSV and SpSM but we can also wait the next release.
> 
> I'd be fine merging now, seeing how there's functionality/tests that are working already. Maybe add a note of open an issue to keep track of the functionality to enable for the next CUSPARSE release though?

Yes, let's do that.
I opened the issue #1610.

File: lib/cusparse/generic.jl [hunk 2]
```diff
@@ -345,3 +359,103 @@ function mm!(transa::SparseChar, transb::SparseChar, α::Number, A::CuSparseMatr
...
+
+    (mA != nA) && throw(DimensionMismatch("A must be square, but has dimensions ($mA,$nA)!"))
+    (mC != mA) && throw(DimensionMismatch("C must have $mA rows, but has $mC rows"))
+    (mB != mA) && (transb == 'N') && throw(DimensionMismatch("B must have $mA rows, but has $mB rows"))
+    (nB != mA) && (transb != 'N') && throw(DimensionMismatch("B must have $mA columns, but has $nB columns"))
+    (nB != nC) && (transb == 'N') && throw(DimensionMismatch("B and C must have the same number of columns, but B has $nB columns and C has $nC columns"))
+    (mB != nC) && (transb != 'N') && throw(DimensionMismatch("B must have the same the number of rows that C has as columns, but B has $mB rows and C has $nC columns"))
+
+    if isa(A, CuSparseMatrixCSC)
+        # cusparseSpSM doesn't supports CSC format
```
Comment (amontoison) [inline lib/cusparse/generic.jl]:
```suggestion
        # cusparseSpSM doesn't support CSC format
```

File: lib/cusparse/generic.jl [hunk 3]
```diff
@@ -345,3 +359,103 @@ function mm!(transa::SparseChar, transb::SparseChar, α::Number, A::CuSparseMatr
...
+    mA,nA = size(A)
+    mX = length(X)
+    mY = length(Y)
+
+    (mA != nA) && throw(DimensionMismatch("A must be square, but has dimensions ($mA,$nA)!"))
+    (mX != mA) && throw(DimensionMismatch("X must have length $mA, but has length $mX"))
+    (mY != mA) && throw(DimensionMismatch("Y must have length $mA, but has length $mY"))
+
+    if isa(A, CuSparseMatrixCSC)
+        # cusparseSpSV doesn't supports CSC format
```
Comment (amontoison) [inline lib/cusparse/generic.jl]:
```suggestion
        # cusparseSpSV doesn't support CSC format
```

**Comment by maleadt** [comment]
CI looks good, let's merge this. Thanks for the PR!

For future reference though, it may be better to keep the PR as 'draft' for as long as possible, to save on CI time when it is known not to be ready.

**Pull request merged**